neo4j/work/simple.py

```diff
@@ -675,24 +758,39 @@ def keys(self):
             return self._metadata.get("fields")
 
     def records(self):
-        """ Generator for records obtained from this result.
+        """Generator for records obtained from this result.
 
         :yields: iterable of :class:`.Record` objects
         """
-        records = self._records
-        next_record = records.popleft
-        while records:
+        records_buffer = self._records
+        next_record = records_buffer.popleft
+        while records_buffer:
             yield next_record()
         attached = self.attached
         if attached():
             self._session.send()
         while attached():
-            self._session.fetch()
-            while records:
-                yield next_record()
-
-    def summary(self):
-        """ Obtain the summary of this result, buffering any remaining records.
+            _ = self._session.fetch()  # Blocking call, this call can detach the session
+            if self._session:
```

Comment:
This looks weird, we touched self._session on the line above.
But I guess this is related to circular references between result and session, still weird programming model or maybe a type?